Framework/Core/include/Framework/ContextRegistry.h

```diff
@@ -11,55 +11,63 @@
 #ifndef FRAMEWORK_CONTEXTREGISTRY_H
 #define FRAMEWORK_CONTEXTREGISTRY_H
 
-#include <array>
+#include <unordered_map>
+#include <typeinfo>
+#include <typeindex>
+#include <type_traits>
+#include <string>
+#include <stdexcept>
 
 namespace o2
 {
 namespace framework
 {
 
+/// @class ContextRegistry
+/// Instances are registered by pointer and are not owned by the registry
 /// Decouples getting the various contextes from the actual type
 /// of context, so that the DataAllocator does not need to know
-/// about the various serialization methods. Since there is only
-/// a few context types we support, this can be done in an ad hoc
-/// manner making sure each overload of ContextRegistry<T>::get()
-/// uses a different entry in ContextRegistry::contextes;
+/// about the various serialization methods.
 ///
-/// Right now we use:
-///
-/// MessageContext 0
-/// ROOTObjectContext 1
-/// StringContext 2
-/// ArrowContext 3
-/// RawContext 4
 class ContextRegistry
 {
  public:
-  ContextRegistry(std::array<void*, 5> contextes)
-    : mContextes{ contextes }
+  ContextRegistry();
+
+  template <typename... Types>
+  ContextRegistry(Types*... instances)
+  {
+    set(std::forward<Types*>(instances)...);
+  }
+
+  template <typename T>
+  T* get() const
   {
+    void* instance = nullptr;
+    try {
+      instance = mRegistry.at(std::type_index(typeid(T)));
```

Comment:
Actually I think you can use `typeid(T*)` to make it faster, since it's guaranteed not to be polymorphic and then `typeid` should be a constexpr.

Reply:
ok, trying this also with the benchmark

Reply:
typeid returns type_info which can not be used as key directly, but there is type_info.hash_code() and it indeed makes the thing faster.

Reply:
I meant `std::type_index(typeid(T*))`. What numbers do you get?